### PR TITLE
hevm: implement subroutines

### PR DESCRIPTION
**Pull request by xwvvvvwx**
Implements subroutines based on the spec at: https://eips.ethereum.org/EIPS/eip-2315

This code passes all the tests here: https://github.com/ethereum/tests/tree/develop/BlockchainTests/GeneralStateTests/stSubroutine

For some reason I couldn't get `hevm compliance` to run these tests (they work fine with `hevm bc-test` however....), so I didn't update the commit for `ethereum/tests` in `release.nix` yet @MrChico @livnev do you have any ideas?

I'm also wondering if we want to add some visualization of the current state of the `return_stack` to `hevm` in some future PR?

I’m also not sure if it makes sense to reuse the `BadJumpDestination` and `StackLimitExceeded` errors or make new ones for the subroutines context?

### CODE REVIEW

File: src/hevm/src/EVM.hs [hunk 1]
```diff
@@ -906,6 +910,33 @@ exec1 = do
...
+
+        -- op: JUMPSUB
+        0x5e ->
+          case stk of
+            (x:xs) ->
+              burn g_high $ do
+                theCode <- use (state . code)
+                theReturnStack <- use (state . return_stack)
+                thePc <- use (state . pc)
+                if x < num (BS.length theCode) && BS.index theCode (num x) == 0x5c
```
Comment (MrChico) [inline src/hevm/src/EVM.hs:931]: Don't you need to make sure that this `0x5c` is not inside a push? Cf. the `checkJump` function and the `codeOps` checking

**Comment by xwvvvvwx** [comment]
CI failed due to #395 :cry:

File: src/hevm/src/EVM.hs (hunk 1 above)
Comment (xwvvvvwx) [reply src/hevm/src/EVM.hs:931]: aha! It was not at all clear to me what the `codeOps` check was there for. Thx :hugs:

**Comment by MrChico** [comment]
there are some other proposed changes to this EIP that have been discussed [here](https://ethereum-magicians.org/t/eip-2315-simple-subroutines-for-the-evm-analysis/4229), not sure what the current status is

**Comment by xwvvvvwx** [comment]
rebased on `master`

**Comment by xwvvvvwx** [comment]
> there are some other proposed changes to this EIP that have been discussed [here](https://ethereum-magicians.org/t/eip-2315-simple-subroutines-for-the-evm-analysis/4229), not sure what the current status is

Doesn't look they're going in afaict: https://github.com/ethereum/EIPs/pull/2663

File: src/hevm/src/EVM.hs (hunk 1 above)
Comment (xwvvvvwx) [reply src/hevm/src/EVM.hs:929]: Is there a reason why `checkJump` checks for `0x5b` in `theCode` and also for `OpJumpdest` in `theCodeOps`? 

Afaict the first check (in `theCode`) is not required if the second (in `theCodeOps`) is present?

Comment (MrChico) [reply src/hevm/src/EVM.hs:929]: yeah, you're right. Might be an optimization, but I'm not sure

Comment (xwvvvvwx) [reply src/hevm/src/EVM.hs]: `JUMPSUB` should now reject jumps to pushdata. I also factored the `theCodeOps` lookup into a helper and removed the duplicated check in `checkJump`.

File: src/hevm/src/EVM.hs [hunk 2]
```diff
@@ -904,6 +908,36 @@ exec1 = do
...
+                theCode <- use (state . code)
+                theCodeOps <- use (env . contracts . ix self . codeOps)
+                theReturnStack <- use (state . return_stack)
+                thePc <- use (state . pc)
+                if x < num (BS.length theCode) then
+                  case findOpAt x OpBeginsub theCodeOps of
+                    Nothing ->  vmError BadJumpDestination
+                    _ -> if length theReturnStack < 1023 then do
+                           state . stack .= xs
+                           state . return_stack %= (thePc + 1 :)
```
Comment (MrChico) [inline src/hevm/src/EVM.hs:935]:
```suggestion
                           pushTo (state . return_stack) (thePc + 1)
```

**Comment by xwvvvvwx** [comment]
subroutines are dropped from berlin: https://github.com/ethereum/pm/issues/263

**Pull request closed without merging**